### PR TITLE
Feature lwdev 6659 lykke pay

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.Service.EthereumCore.AzureRepositories/Repositories/BlockSyncedByHashRepository.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        public async Task DeleteByPartitionAndHashAsync(string partition, string hash)
+        {
+            await _table.DeleteIfExistAsync(partition, hash?.ToLower());
+        }
+
+        public async Task<IBlockSyncedByHash> GetLastSyncedAsync(string partition)
+        {
+            var index = await _index.GetDataAsync(_lastSyncedPartition, partition);
+            var entity = await _table.GetDataAsync(index);
```
Comment (tarurar) [inline src/Lykke.Service.EthereumCore.AzureRepositories/Repositories/BlockSyncedByHashRepository.cs]: Check index for null before passing it to GetDataAsync

File: src/Lykke.Service.EthereumCore.RabbitMQ/RegisterRabbitQueue.cs [hunk 2]
```diff
@@ -33,7 +36,7 @@ public static class RegisterRabbitQueueEx
                 ExchangeName = exchangeName,
                 DeadLetterExchangeName = $"{exchangeName}.dlx",
                 RoutingKey = ""
-//                IsDurable = true
+                //                IsDurable = true
```
Comment (tarurar) [inline src/Lykke.Service.EthereumCore.RabbitMQ/RegisterRabbitQueue.cs]: Why not remove the line?

File: src/Lykke.Service.EthereumCore/Controllers/LykkePay/LykkePayErc20DepositContractsController.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            _contractService = contractService;
+        }
+
+        [HttpPost]
+        [ProducesResponseType(typeof(RegisterResponse), 200)]
+        [ProducesResponseType(typeof(ApiException), 400)]
+        [ProducesResponseType(typeof(void), 401)]
+        [ProducesResponseType(typeof(ApiException), 500)]
+        public async Task<IActionResult> CreateDepositContractAsync([FromQuery] string userAddress)
```
Comment (tarurar) [inline src/Lykke.Service.EthereumCore/Controllers/LykkePay/LykkePayErc20DepositContractsController.cs]: What if userAddress is empty or null?

File: src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using MoreLinq;
+using Module = Autofac.Module;
+
+namespace Lykke.Service.EthereumCore.Modules
+{
+    public class ControllerModule : Module
+    {
+        private readonly ILog _log;
+        //private readonly IServiceCollection _services;
```
Comment (tarurar) [inline src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs]: Please remove comments

File: src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+using Lykke.Service.RabbitMQ;
+using Lykke.SettingsReader;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using MoreLinq;
+using Module = Autofac.Module;
+
+namespace Lykke.Service.EthereumCore.Modules
+{
+    public class ControllerModule : Module
```
Comment (tarurar) [inline src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs]: As far as I see the purpose of this module is to register Controllers but there is no need in registering controllers manually.

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+using Lykke.Service.EthereumCore.Core.Shared;
+using Nethereum.Contracts;
+using Nethereum.Hex.HexTypes;
+
+namespace Lykke.Service.EthereumCore.Services.LykkePay
+{
+    //LykkePay Erc20 deposit contract service
+    public class LykkePayErc20DepositContractService : IErc20DepositContractService
+    {
+        /*
```
Comment (tarurar) [inline src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: Please remove comments

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+            var guid = Guid.NewGuid();
+            var guidStr = guid.ToString();
+
+            var message = new Lykke.Service.EthereumCore.Core.Messages.LykkePay.LykkePayErc20TransferMessage()
+            {
+                OperationId = guidStr
+            };
+
+            var existingOperation = await _operationsRepository.GetAsync(guid.ToString());
```
Comment (tarurar) [inline src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: You have already defined guidStr variable earlier, use it.

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+                throw new ClientSideException(ExceptionType.TransferInProcessing, $"Transfer from {depositContractAddress} was started before wait for it to complete");
+            }
+
+            var balance = await _ercInterfaceService.GetBalanceForExternalTokenAsync(depositContractAddress, erc20TokenAddress);
+            if (balance == 0)
+            {
+                throw new ClientSideException(ExceptionType.NotEnoughFunds, $"No tokens detected at deposit address {depositContractAddress}");
+            }
+
+            var guid = Guid.NewGuid();
```
Comment (tarurar) [inline src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: As far as I see there is no need in separate guid variable. Please remove it.

File: src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs (hunk 5 above)
Comment (OttoVT) [reply src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs]: You have to register controllers manually if you want to use Named and Keyed Services all across your app (Autofac feature)

File: src/Lykke.Job.EthereumCore/Job/Erc20DepositContractPoolJob.cs [hunk 9]
```diff
@@ -32,5 +39,19 @@ public class Erc20DepositContractPoolJob
...
             }
         }
+
+        [TimerTrigger("0.00:01:00")]
+        public async Task ExecuteForLykkeApi()
+        {
+            try
+            {
+                await _contractPoolServiceLykkePay.ReplenishPool();
+                await _logger.WriteInfoAsync(nameof(Erc20DepositContractPoolJob), nameof(ExecuteForLykkeApi), "", "Pool have been replenished", DateTime.UtcNow);
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Lykke.Job.EthereumCore/Job/Erc20DepositContractPoolJob.cs]: +

File: src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayHotWalletMonitoringTransactionJob.cs [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                await _hotWalletCashoutTransactionRepository.GetByOperationIdAsync(operationId);
+            var cashout = await _hotWalletCashoutRepository.GetAsync(cashoutTransaction?.OperationId);
+
+            return cashout;
+        }
+
+        private async Task RepeatOperationTillWin(CoinTransactionMessage message)
+        {
+            var operation = await GetOperationAsync(message?.TransactionHash, message?.OperationId);
+            switch (operation.OperationType)
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayHotWalletMonitoringTransactionJob.cs]: fixed

File: src/Lykke.Job.EthereumCore/Job/Erc20DepositContractPoolJob.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Lykke.Job.EthereumCore/Job/Erc20DepositContractPoolJob.cs]: fixed

File: src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayTransferNotificationJob.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            _logger = logger;
+            _operationsRepository = operationsRepository;
+            _rabbitQueuePublisher = rabbitQueuePublisher;
+            _web3 = web3;
+        }
+
+        [QueueTrigger(Constants.LykkePayErc223TransferNotificationsQueue, 200, true)]
+        public async Task Execute(LykkePayErc20TransferNotificationMessage message, QueueTriggeringContext context)
+        {
+            try
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayTransferNotificationJob.cs]: fixed

File: src/Lykke.Service.EthereumCore.AzureRepositories/Repositories/BlockSyncedByHashRepository.cs (hunk 1 above)
Comment (OttoVT) [reply src/Lykke.Service.EthereumCore.AzureRepositories/Repositories/BlockSyncedByHashRepository.cs]: fixed

File: src/Lykke.Service.EthereumCore.RabbitMQ/RegisterRabbitQueue.cs (hunk 2 above)
Comment (OttoVT) [reply src/Lykke.Service.EthereumCore.RabbitMQ/RegisterRabbitQueue.cs]: fixed

File: src/Lykke.Service.EthereumCore/Controllers/LykkePay/LykkePayErc20DepositContractsController.cs (hunk 3 above)
Comment (OttoVT) [reply src/Lykke.Service.EthereumCore/Controllers/LykkePay/LykkePayErc20DepositContractsController.cs]: fixed

File: src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs (hunk 4 above)
Comment (OttoVT) [reply src/Lykke.Service.EthereumCore/Modules/ControllerModule.cs]: fixed

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs (hunk 6 above)
Comment (OttoVT) [reply src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: fixed

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs (hunk 7 above)
Comment (OttoVT) [reply src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: fixed

File: src/Services/LykkePay/LykkePayErc20DepositContractService.cs (hunk 8 above)
Comment (OttoVT) [reply src/Services/LykkePay/LykkePayErc20DepositContractService.cs]: fixed

File: src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayHotWalletMonitoringTransactionJob.cs [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+                }
+            }
+        }
+
+        private async Task<IHotWalletOperation> GetOperationAsync(string trHash, string operationId)
+        {
+            operationId = operationId ?? "";
+            var cashoutTransaction = await _hotWalletCashoutTransactionRepository.GetByTransactionHashAsync(trHash) ??
+                await _hotWalletCashoutTransactionRepository.GetByOperationIdAsync(operationId);
+            var cashout = await _hotWalletCashoutRepository.GetAsync(cashoutTransaction?.OperationId);
```
[the remark this answers is not in the document]
Comment (OttoVT) [reply src/Lykke.Job.EthereumCore/Job/LykkePay/LykkePayHotWalletMonitoringTransactionJob.cs]: it returns null

**Pull request merged**